pkgs/development/compilers/fbc/default.nix

```diff
@@ -0,0 +1,165 @@
+    })
+    # Respect SOURCE_DATE_EPOCH when set
+    # Remove on next release after PR merge
+    (fetchpatch {
+      name = "fbc-SOURCE_DATE_EPOCH-support.patch";
+      url = "https://github.com/freebasic/fbc/pull/392/commits/74ea6efdcfe9a90d1c860f64d11ab4a6cd607269.patch";
+      excludes = [ "changelog.txt" ];
+      sha256 = "sha256-v5FTi4vKOvSV03kigZDiOH8SEGEphhzkBL6p1hd+NtU=";
+    })
+  ];
+
+  postPatch = ''
+    patchShebangs tests/warnings/test.sh
+
+    # Some tests lack proper dependency on libstdc++
+    for missingStdcpp in tests/cpp/{class,call2}-fbc.bas; do
+      sed -i -e "/'"' TEST_MODE : /a #inclib "stdc++"' $missingStdcpp
+    done
+
+    # Help compiler find libstdc++ with gcc backend
+    sed -i -e '/fbcAddLibPathFor( "libgcc.a" )/a fbcAddLibPathFor( "libstdc++.so" )' src/compiler/fbc.bas
+  '';
+
+  dontConfigure = true;
+
+  depsBuildBuild = [
+    buildPackages.stdenv.cc
+    buildPackages.ncurses
+    buildPackages.libffi
+  ];
+
+  buildInputs = [
+    ncurses
+    libffi
+  ] ++ lib.optionals stdenv.hostPlatform.isLinux [
+    gpm
+    libGL
+    libX11
+    libXext
+    libXpm
+    libXrandr
+  ];
+
+  enableParallelBuilding = true;
+
+  hardeningDisable = [
+    "format"
+  ];
+
+  makeFlags = lib.optionals (stdenv.buildPlatform != stdenv.hostPlatform) [
+    "TARGET=${stdenv.hostPlatform.config}"
+  ];
+
+  preBuild = ''
+    export buildJobs=$NIX_BUILD_CORES
+    if [ -z "$enableParallelBuilding" ]; then
+      buildJobs=1
+    fi
+
+    echo Bootstrap an unpatched build compiler
+    make bootstrap-minimal -j$buildJobs \
+      BUILD_PREFIX=${buildPackages.stdenv.cc.targetPrefix} LD=${buildPackages.stdenv.cc.targetPrefix}ld
+
+    echo Compile patched build compiler and host rtlib
+    make compiler -j$buildJobs \
+      "FBC=$PWD/bin/fbc${stdenv.buildPlatform.extensions.executable} -i $PWD/inc" \
+      BUILD_PREFIX=${buildPackages.stdenv.cc.targetPrefix} LD=${buildPackages.stdenv.cc.targetPrefix}ld
+    make rtlib -j$buildJobs \
+      "FBC=$PWD/bin/fbc${stdenv.buildPlatform.extensions.executable} -i $PWD/inc" \
+      ${if (stdenv.buildPlatform == stdenv.hostPlatform) then
+        "BUILD_PREFIX=${buildPackages.stdenv.cc.targetPrefix} LD=${buildPackages.stdenv.cc.targetPrefix}ld"
+      else
+        "TARGET=${stdenv.hostPlatform.config}"
+      }
+
+    echo Install patched build compiler and host rtlib to local directory
+    make install-compiler prefix=$PWD/patched-fbc
+    make install-rtlib prefix=$PWD/patched-fbc ${lib.optionalString (stdenv.buildPlatform != stdenv.hostPlatform) "TARGET=${stdenv.hostPlatform.config}"}
+    make clean
+
+    echo Compile patched host everything with previous patched stage
+    buildFlagsArray+=("FBC=$PWD/patched-fbc/bin/fbc${stdenv.buildPlatform.extensions.executable} -i $PWD/inc")
+  '';
+
+  installFlags = [
+    "prefix=${placeholder "out"}"
+  ];
+
+  doCheck = stdenv.buildPlatform == stdenv.hostPlatform;
```

Comment:
```suggestion
  doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform;
```

Reply:
I tried `pkgsCross.gnu32.fbc` with this change + passing along `CC` to the test makefiles (otherwise they'd use the buildPlatform one) and one of the `log-tests` suite fails:
```
pretest/multi_module_ok.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/call.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/mangle.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/fbcall.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/bop.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/derived.bmk : TEST_MODE=MULTI_MODULE_OK
make[2]: *** [log-tests.mk:259: cpp/derived.log] Error 1
```

`cpp/derived.log`:
```
cpp/derived.bmk : TEST_MODE=MULTI_MODULE_OK
make[3]: Entering directory '/build/source/tests'
/build/source/tests/../bin/fbc -i /build/source/tests/../inc -lang fb -target i686-unknown-linux-gnu -g -w 0  -m cpp/derived-fbc -c cpp/derived-fbc.bas -o cpp/derived-fbc.o
# Pass -m32 to get -m32 or -m64 as required
i686-unknown-linux-gnu-g++ -c -m32 -Wall -Wno-attributes -o cpp/derived-cpp.o cpp/derived-cpp.cpp
/build/source/tests/../bin/fbc -i /build/source/tests/../inc -target i686-unknown-linux-gnu -g cpp/derived-fbc.o cpp/derived-cpp.o -x cpp/derived-fbc
cpp/derived-fbc: error while loading shared libraries: libgcc_s.so.1: cannot open shared object file: No such file or directory
make[3]: *** [bmk-make.mk:137: run_test_ok] Error 1
make[3]: Leaving directory '/build/source/tests'
cpp/derived.bmk : RESULT=FAILED
```

```
↪ file /tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc
/tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib/ld-linux.so.2, for GNU/Linux 2.6.32, not stripped

↪ nix-shell -p patchelf --run 'patchelf --print-rpath /tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc'
/nix/store/b5pm60096f94nh23c5xlph0690338cjj-fbc-i686-unknown-linux-gnu-1.09.0/lib64:/nix/store/b5pm60096f94nh23c5xlph0690338cjj-fbc-i686-unknown-linux-gnu-1.09.0/lib:/nix/store/3dvhlw5kbm3l8j6567l2n4h4db5hqg7k-i686-unknown-linux-gnu-stage-final-gcc-10.4.0-lib/i686-unknown-linux-gnu/lib:/nix/store/6crjr3xsx99fkg0hxcwhyawbrycj8k9j-ncurses-i686-unknown-linux-gnu-6.3-p20220507/lib:/nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib

↪ find /nix/store/3dvhlw5kbm3l8j6567l2n4h4db5hqg7k-i686-unknown-linux-gnu-stage-final-gcc-10.4.0-lib/ -name 'libgcc_s.so.1'
/nix/store/3dvhlw5kbm3l8j6567l2n4h4db5hqg7k-i686-unknown-linux-gnu-stage-final-gcc-10.4.0-lib/i686-unknown-linux-gnu/lib/libgcc_s.so.1

↪ strace /tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc 2>&1 | grep libgcc
openat(AT_FDCWD, "/nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib/libgcc_s.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib/libgcc_s.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
writev(2, [{iov_base="/tmp/nix-build-fbc-i686-unknown-"..., iov_len=83}, {iov_base=": ", iov_len=2}, {iov_base="error while loading shared libra"..., iov_len=36}, {iov_base=": ", iov_len=2}, {iov_base="libgcc_s.so.1", iov_len=13}, {iov_base=": ", iov_len=2}, {iov_base="cannot open shared object file", iov_len=30}, {iov_base=": ", iov_len=2}, {iov_base="No such file or directory", iov_len=25}, {iov_base="\n", iov_len=1}], 10/tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc: error while loading shared libraries: libgcc_s.so.1: cannot open shared object file: No such file or directory
```

Not sure what to make of this. `pkgsi686Linux.fbc` itself works though.

```
↪ file /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc
/tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /nix/store/n2d7w74v3ixjsfd96b3207nzj6w4rl1f-glibc-2.35-163/lib/ld-linux.so.2, for GNU/Linux 2.6.32, not stripped

↪ nix-shell -p patchelf --run 'patchelf --print-rpath /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc'
/nix/store/7kg5b400f3g5shkvkb647kwv5m2mnmb1-fbc-1.09.0/lib:/nix/store/dk0f8ffm2zyf4q3gqn035n13znhpdzmk-gcc-10.4.0-lib/lib:/nix/store/dfn02wyp7j76aw3f20mcbmar19siapzw-ncurses-6.3-p20220507/lib:/nix/store/n2d7w74v3ixjsfd96b3207nzj6w4rl1f-glibc-2.35-163/lib

↪ strace /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc 2>&1 | grep libgcc
openat(AT_FDCWD, "/nix/store/dk0f8ffm2zyf4q3gqn035n13znhpdzmk-gcc-10.4.0-lib/lib/libgcc_s.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
```

Reply:
I've added a comment explaining that tests do not pass under cross, I think that's all I can do here.

Reply:
> If that's the new recommended way of doing it then the manual should get updated.

Depends a bit on the usecase. It does not always work but eg when generating man pages it should be used and we should probably update the docs to consider this.